src/pages/api/stripe/webhook.ts

```diff
@@ -11,6 +11,21 @@ function stripeToMixpanelDataConverter(stripeDate: number) {
   return date.toISOString()
 }
 
+type PlanInterval = 'month' | 'year' | 'quarter'
+
+function checkForUpgrade(
+  previousPlanInterval: PlanInterval,
+  newPlanInterval: PlanInterval,
+) {
+  const previousPlanIntervalUpgrades = {
+    month: true,
+    year: false,
+    quarter: newPlanInterval === 'year',
+  }
+
+  return previousPlanIntervalUpgrades[previousPlanInterval] ?? false
```

Comment:
Multiple if checks can be turned into a data structure for cleaner code. Thanks @theianjones

Reply:
This is a bit more verbose, but I think it gives more information and ratchets up the type safety. I'm on a bit of Zod kick and I think it is well suited for this scenario, where we have unknown/uncontrolled outside input coming in from a 3rd party webhook.

This is what the `checkForUpgrade` function could look like along with types.

```typescript
import {z} from 'zod';

const UPGRADE = "Upgrade";
const DOWNGRADE = "Downgrade";

const intervalOptions = z.union([
  z.literal("month"),
  z.literal("quarter"),
  z.literal("year")
]);

type Interval = z.infer<typeof intervalOptions>;
type IntervalChange = typeof UPGRADE | typeof DOWNGRADE;

function checkForUpgrade(
  previousPlanInterval: Interval,
  newPlanInterval: Interval
): IntervalChange {
  const isAnUpgrade = {
    month: true,
    year: false,
    quarter: newPlanInterval === "year"
  }[previousPlanInterval];

  return isAnUpgrade ? UPGRADE : DOWNGRADE;
}
```

And then the webhook code that checks the upgrade type would look something like:

```typescript
const previousPlanInterval = previousSubscription.plan?.interval
const newPlanInterval = stripeSubscription.items.data[0].plan?.interval

const intervalTuple = z.tuple([intervalOptions, intervalOptions]);
const result = intervalTuple.safeParse([
  previousPlanInterval,
  newPlanInterval
]);

if(!result.success) {
  const errorMessage = `
  Invalid interval types ahead of checkForUpgrade
  - previousPlanInterval: ${previousPlanInterval}
  - newPlanInterval: ${newPlanInterval}

  Detailed ZodError: ${result.error}
  `

  // OR, if we don't want to throw an error, we could send the message to something
  // like Honeybadger and then early return from the webhook handler.
  throw new Error(errorMessage)
}

// because of the `intervalTupe.safeParse(...)` above, we know we are passing in
// valid inputs to the checkForUpgrade
const state = checkForUpgrade(...result.data);
```

That's it for the code suggestion.

As a side note, I ran through the different possibilities to double check with you that this function produces the results you want.

```typescript
const options = intervalOptions.options;
const values = options.map((option) => option.value);

for (const previousInterval of values) {
  for (const newInterval of values) {
    const result = checkForUpgrade(
      previousInterval as Interval,
      newInterval as Interval
    );
    console.log(`- ${previousInterval} | ${newInterval} --> ${result}`);
  }
}
```

which spits out:

```
- month | month --> Upgrade 
- month | quarter --> Upgrade 
- month | year --> Upgrade 
- quarter | month --> Downgrade 
- quarter | quarter --> Downgrade 
- quarter | year --> Upgrade 
- year | month --> Downgrade 
- year | quarter --> Downgrade 
- year | year --> Downgrade 
```

Reply:
also, these are just suggestions, no pressure if you're wanting to get out the fix and not get caught up in this kind of refactor

Reply:
Love this, I'm going to reach for Zod when I have heavier-duty typing to do